src/ingest/src/main/scala/osmesa/ingest/ProcessOSM.scala

```diff
@@ -304,6 +308,141 @@ object ProcessOSM {
       .where('visible and isnull('validUntil)) // This filters things down to all and only the most current geoms which are visible
   }
 
+//  def reconstructRelationGeometries(ppnodes: DataFrame, wayGeoms: DataFrame, pprelations: DataFrame): DataFrame = ???
+
+  // create fully-formed rings from line segments
+  @tailrec
+  private def connectSegments(segments: List[Line], rings: List[Line] = List.empty[Line]): List[Line] = {
+    segments match {
+      case Nil => rings
+      case h :: t if h.isClosed => connectSegments(t, rings :+ h)
+      case h :: t =>
+        connectSegments(t.find(line => h.vertices.last == line.vertices.head) match {
+          case Some(next) => Line(h.vertices ++ next.vertices.tail) :: t.filterNot(line => line == next)
+          case None =>
+            t.find(line => h.vertices.last == line.vertices.last) match {
+              case Some(next) => Line(h.vertices ++ next.vertices.reverse.tail) :: t.filterNot(line => line == next)
+              case None => throw new Exception("Unable to connect segments.")
+            }
+        }, rings)
+    }
+  }
+
+  @tailrec
+  private def dissolveRings(rings: List[Polygon], dissolvedOuters: List[Line] = List.empty[Line], dissolvedInners: List[Line] = List.empty[Line]): (List[Line], List[Line]) = {
+    rings match {
+      case Nil => (dissolvedOuters, dissolvedInners)
+      case h :: t =>
+        val touching = t.filter(r => h.touches(r))
+        val remaining = t.filterNot(r => h.touches(r))
+
+        touching match {
+          case _ if touching.isEmpty => dissolveRings(t.filterNot(r => h.touches(r)), dissolvedOuters :+ h.exterior, dissolvedInners ++ h.holes)
+          case _ =>
+            val dissolved = touching.foldLeft(List(h)) {
+              case (rs, r2) =>
+                rs.flatMap { r =>
+                  r.union(r2).toGeometry match {
+                    case Some(p: Polygon) => List(p)
+                    case Some(mp: MultiPolygon) => mp.polygons
+                    case _ => throw new Exception("Union failed.")
+                  }
+                }
+            }
+
+            val retryRings = dissolved.filter(d => remaining.exists(r => r.touches(d)))
+            val newRings = dissolved.filter(d => !remaining.exists(r => r.touches(d)))
+
+            dissolveRings(retryRings ++ remaining, dissolvedOuters ++ newRings.map(_.exterior), dissolvedInners ++ newRings.flatMap(_.holes))
+        }
+    }
+  }
+
+  private def dissolveRings(rings: List[Line]): (List[Line], List[Line]) = {
+    dissolveRings(rings.map(x => Polygon(x)))
+  }
+
+  val buildMultiPolygon: UserDefinedFunction = udf((id: Long, ways: Seq[Row]) => {
+    try {
+      val coords: Seq[(String, Line)] = ways
+        .map(row => (row.getAs[String]("role"), row.getAs[Array[Byte]]("geom").readWKB match {
+          case geom: Polygon => geom.as[Polygon].map(_.exterior)
+          case geom => geom.as[Line]
```

Comment:
These should probably get converted into `Polygon`s here to avoid creating them multiple times below, yeah?